pkg/controller/controller_instance.go

```diff
@@ -329,6 +367,165 @@ func (c *controller) initOrphanMitigationCondition(instance *v1beta1.ServiceInst
+// namespace/instancename@generation
+func (c *controller) getInstanceAndGenerationAsKey(instance *v1beta1.ServiceInstance) (string, error) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(instance)
+	if err != nil {
+		err := fmt.Errorf("Couldn't get key for object %+v: %v", instance, err)
+		return "", err
+	}
+
+	key = strings.Join([]string{key, "@", strconv.FormatInt(instance.Generation, 10)}, "")
+	return key, nil
+}
+
+// setRetryBackoffRequired marks the specified instance/generation as needing a
+// delay before the next provision/update is attempted.  We always set this flag
+// before attempting an update or delete operation in case we must retry.  This
+// will eventually be cleared by background worker running purgeExpiredRetryEntries()
+// or when the operation is successful.
+func (c *controller) setRetryBackoffRequired(instance *v1beta1.ServiceInstance) {
+	key, err := c.getInstanceAndGenerationAsKey(instance)
+	if err != nil {
+		glog.Errorf(err.Error())
+		return
+	}
+
+	c.instanceOperationRetryQueue.mutex.Lock()
+	defer c.instanceOperationRetryQueue.mutex.Unlock()
+	// older generations are meaningless, purge any
+	deleteInstanceKeysFromMapForAllGens(c.instanceOperationRetryQueue.backoffBeforeRetrying, key)
+	c.instanceOperationRetryQueue.backoffBeforeRetrying[key] = time.Now()
+	glog.V(4).Infof("added %v to backoffBeforeRetrying", key)
+}
+
+// deleteInstanceKeysFromMapForAllGens purges out any instance entries ignoring the generation
+// Assumes the caller did the neccessary mutex locking for thread safety
+func deleteInstanceKeysFromMapForAllGens(m map[string]time.Time, key string) {
+	end := strings.Index(key, "@")
+	if end > 1 {
+		anyGeneration := key[0:end]
+		for k := range m {
+			if strings.HasPrefix(k, anyGeneration) {
+				delete(m, k)
+			}
+		}
+	}
+}
+
+// backoffAndRequeueIfRetrying returns true if this is a retry and a backoff
+// (delay) needs to be observed before retrying.  This only applies to
+// Provisioning and Updating and is generation specific.  If the generation has
+// been bumped since the instance was added to the retry map there will be no
+// backoff delay.  Check if the instance has been marked as needing a delay, if
+// so, determine exponential backoff, calculate next retry time, and clear from
+// the backoffBeforeRetrying queue
+func (c *controller) backoffAndRequeueIfRetrying(instance *v1beta1.ServiceInstance, operation string) bool {
+	key, err := c.getInstanceAndGenerationAsKey(instance)
+	if err != nil {
+		glog.Error(err.Error())
+		return false
+	}
+	delay := time.Millisecond * 0
+
+	// if there is a pending delay, calculate it and clear it
+	c.instanceOperationRetryQueue.mutex.RLock()
+	_, exists := c.instanceOperationRetryQueue.backoffBeforeRetrying[key]
+	c.instanceOperationRetryQueue.mutex.RUnlock()
+
+	if exists {
+		c.instanceOperationRetryQueue.mutex.Lock()
+		if _, exists := c.instanceOperationRetryQueue.backoffBeforeRetrying[key]; exists {
```

Comment:
Can you add a comment here that explains why the check in the `retryTime` queue is not needed in the case where the key has been purged from the `backoffBeforeRetrying` queue since the earlier check in this function?